cdi/src/main/java/io/smallrye/concurrency/cdi/context/propagation/CdiContextProvider.java

```diff
@@ -0,0 +1,136 @@
+            if (scopeToContextualInstances.containsKey(requestContext.getScope())) {
+                requestMap = new HashMap<>();
+                requestContext.associate(requestMap);
+                requestContext.activate();
+            }
+            if (scopeToContextualInstances.containsKey(sessionContext.getScope())) {
+                sessionMap = new HashMap<>();
+                sessionContext.associate(sessionMap);
+                sessionContext.activate();
+            }
+            if (scopeToContextualInstances.containsKey(conversationContext.getScope())) {
+                // for conversation scope, we need request and session storages, only proceed when that condition is met
+                if (requestMap != null && sessionMap != null) {
+                    conversationContext.associate(new MutableBoundRequest(requestMap, sessionMap));
+                    conversationContext.activate();
+                }
+            }
+
+            // propagate all contexts that have some bean in them
+            if (scopeToContextualInstances.get(requestContext.getScope()) != null) {
+                requestContext.clearAndSet(scopeToContextualInstances.get(requestContext.getScope()));
+            }
+            if (scopeToContextualInstances.get(sessionContext.getScope()) != null) {
+                sessionContext.clearAndSet(scopeToContextualInstances.get(sessionContext.getScope()));
+            }
+            if (scopeToContextualInstances.get(conversationContext.getScope()) != null) {
+                conversationContext.clearAndSet(scopeToContextualInstances.get(conversationContext.getScope()));
+            }
+
+            // return ThreadContextController which should revert state to before propagation - e.g. clear CDI contexts
+            ThreadContextController controller = () -> {
+                // clean up contexts we previously activated by calling deactivate() on them
+                if (scopeToContextualInstances.containsKey(requestContext.getScope())) {
+                    requestContext.deactivate();
+                }
+                if (scopeToContextualInstances.containsKey(sessionContext.getScope())) {
+                    sessionContext.deactivate();
+                }
+                if (scopeToContextualInstances.containsKey(conversationContext.getScope())) {
+                    conversationContext.deactivate();
+                }
+            };
+
+            // when all is done, return the controller
+            return controller;
+        };
+    }
+
+    @Override
+    public ThreadContextSnapshot clearedContext(Map<String, String> map) {
+        if (!isCdiAvailable()) {
+            //return null as per docs to state that propagation of this context is not supported
+            return null;
+        }
+
+        // by default Application and Singleton scopes propagate anyway
+        // Request, Session and Conversation are not active in any new thread, so no operation should be taken here
```

Comment:
Remember that you cannot rely on having a new thread: it's frequent for callback operations to be executed in the thread that fired the event, which may have a CDI context active that is not the one from the callback operation.